### DISCUSSION TITLE
Has anyone tried dumping a Game Boy Camera save?

### DISCUSSION THREAD

**Discussion by MikalWhaley** [sanni/cartreader#301] [Coding Corner]
I haven't been able to get this to work. I cleaned the contacts on the cart and can get it to read first time on my actual Game Boy but the cart reader shows nothing for the name and an error for the checksum. I read around that the cart is a little different than some of the other GB ones and I'm not sure if that meant SRAM. Even without the name I can dump the file, but the .sav is empty.

**Comment by sanni** [comment]
The GB Camera seems to have an additional chip between the data and the cart edge so the current functions probably need to be slowed down.

You can try doubling the nops starting from Line 432 in GB.ino until the name is read correctly.
https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L432

**Reply by MikalWhaley** [reply]
Hey! Super new to the whole cart dumper thing. Do you mean just doubling the function that's called on line 436 for everything between line 432 and 523? After that I would just flash the file to the cart reader right?

**Comment by sanni** [comment]
Okay. So there are three ways of programming:
- knowing what you doing
- trial&error
- copy&paste

Let's skip over the first option since nobody got time for that. 
The second option was what I suggested in the previous post, just try out some changes to the code and see if it makes a difference, repeat until something changes. This is how I wrote most of the Cart Readers code to be honest.
The smart way is of course looking at how others have done it so let's choose this path for now.

Here is a pretty good write-up about dumping saves from the Game Boy Camera: https://dragaosemchama.com/en/2016/02/downloading-game-boy-camera-pics/

If you have a look at this persons source code and compare it with mine you will notice two major things.
His code:

````
void WriteByte(uint16_t Adr,uint8_t data){
	
	DDRC=0xFF; 
	Go2ADR(Adr);
	DATAOUT=data;
	ControlPort|=(1<<RD);
	ControlPort&=~(1<<WR);
	_delay_us(2);
	ControlPort|=(1<<WR);
		
}
uint8_t readByte(uint16_t Adr){
	
	uint8_t b;
	//DATAOUT=0xFF;////////////////////////////////////////Noobada
	DDRC=0x00;
	Go2ADR(Adr);
	//ControlPort&=~(1<<MREQ);
	ControlPort|=(1<<WR);
	ControlPort&=~(1<<RD);
	_delay_us(2);
	b=DATAIN;
	ControlPort|=(1<<RD);
	//ControlPort|=(1<<MREQ);
	return b;
}
````

My code:
````
byte readByte_GB(word myAddress) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Switch RD(PH6) to LOW
  PORTH &= ~(1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Read
  byte tempByte = PINC;

  // Switch and RD(PH6) to HIGH
  PORTH |= (1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  return tempByte;
}

void writeByte_GB(int myAddress, uint8_t myData) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;
  PORTC = myData;

  // Arduino running at 16Mhz -> one nop = 62.5ns
  // Wait till output is stable
  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull WR(PH5) low
  PORTH &= ~(1 << 5);

  // Leave WE low for at least 60ns
  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull WR(PH5) HIGH
  PORTH |= (1 << 5);

  // Leave WE high for at least 50ns
  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");
}

byte readByteSRAM_GB(word myAddress) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull CS(PH3) LOW
  PORTH &= ~(1 << 3);
  // Pull RD(PH6) LOW
  PORTH &= ~(1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Read
  byte tempByte = PINC;

  // Pull CS(PH3) HIGH
  PORTH |= (1 << 3);
  // Pull RD(PH6) HIGH
  PORTH |=  (1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  return tempByte;
}

void writeByteSRAM_GB(int myAddress, uint8_t myData) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;
  PORTC = myData;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull CS(PH3) LOW
  PORTH &= ~(1 << 3);
  // Pull WR(PH5) low
  PORTH &= ~(1 << 5);

  // Leave WE low for at least 60ns
  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull WR(PH5) HIGH
  PORTH |= (1 << 5);
  // Pull CS(PH3) HIGH
  PORTH |= (1 << 3);

  // Leave WE high for at least 50ns
  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");
}
````

He uses "_delay_us(2)" before reading data which is a 2000ns delay.
I'm using " __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");" which is only 250ns delay.
So we can conclude that my code reads the Game Boy Camera too fast.

What I meant with my previous post was to replace every  "__asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");" with "__asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");" 
So basically doubling the amount of wait time. And if that still doesn't work doubling it again. "__asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");" 


The second difference is that he in his code only toggles CS/MREQ every bank instead of every byte like me

````
if (MBC==3){
		//WriteByte(0x0000, 0x0A);//enable RAM writing
		for (uint8_t i=0;i<0x0F;i++){
			WriteByte(0x4000,i);//select bank
			ControlPort&=~(1<<MREQ);//inside the loop
			for (uint16_t j=0xA000;j<=0xBFFF;j++){
				uart_putc(readByte(j));
			}
			ControlPort|=(1<<MREQ);	
		}
		//WriteByte(0x0000,0x00);//disable RAM writing
	}	
````

````
// Switch SRAM banks
    for (uint8_t bank = 0; bank < sramBanks; bank++) {
      dataOut();
      writeByte_GB(0x4000, bank);

      // Read SRAM
      dataIn_GB();
      for (uint16_t sramAddress = 0xA000; sramAddress <= sramEndAddress; sramAddress += 64) {
        uint8_t readData[64];
        for (uint8_t i = 0; i < 64; i++) {
          readData[i] = readByteSRAM_GB(sramAddress + i);
        }
        myFile.write(readData, 64);
      }
    }
````

But I would try out changing the delay like shown above first. The goal should be that the Cart Reader displays the name of the cartridge correctly "GAMEBOYCAMERA" and also give the correct ROM and RAM size, which according to the link is 64 and 16 banks.

**Reply by MikalWhaley** [reply]
Ah I see that makes sense. I was going to call the entire asm function again instead of just double the nops. I'll definitely play around with this tonight! Thanks for the reply.

**Comment by MikalWhaley** [comment]
After reading your comments in the GB.ino file about how one nop is 62.5ns I added nops that would total 32 for each call as that would be exactly 2000ns and flashed that but no luck. I also added an extra nop for good measure before the next flash with the same result. I would assume this leaves what you suspected at the end of your last post being the CS/MREQ toggles every bank instead of every byte. I'll try to take a stab at that soon next.

**Comment by sanni** [comment]
Here is another code example, this one does a 5ms delay instead after the bank switch. That's a huge delay but it's probably there for a reason.
https://github.com/insidegadgets/GBxCart-RW/blob/master/Interface_Programs/GBxCart_RW_GBCamera_Saver_v1.10/gbxcart_rw_gbcamera_saver_v1.10.c#L176
So you could try putting delay_ms(5); into line 764 https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L764

**Comment by MikalWhaley** [comment]
I kept the added 32 nops as well as adding a _delay_ms(5); (appears to be the equivalent for delay_ms(); ) on to line 764 line but with the same result. I'll take another look and try to discern some more differences between the gbxcart_rw code and this one in the mean time.

**Comment by sanni** [comment]
I noticed two more differences, in GBxCartRW's read_8bit_data function https://github.com/insidegadgets/GBxCart-RW/blob/master/MCU_Files/GBxCart_RW_v1.3_R22/setup.c#L263 he pulls rd high then cs high, while I do the opposite order in my two read functions, might be worth changing the order around in my code:
https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L432
https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L476

The second difference is that I only toggle CS in the readsram function and not in the normal read function, so it might be worth a shot making both function the same and adding the CS toggle to the normal read function.

**Comment by Dakkaron** [comment]
If you just want to get the pictures off the cartridge and don't actually care about the full save, you can use this project: https://github.com/mofosyne/arduino-gameboy-printer-emulator

It's a Gameboy Printer Emulator, using any old Arduino. It emulates a Gameboy Printer towards the Gameboy and turns it into a serial-over-USB stream, which you can then convert into a png using their web page.

I didn't want to cut up a Link Cable, so I bought this Link Cable breakout board as well: https://github.com/Palmr/gb-link-cable

**Comment by MikalWhaley** [comment]
Hello! Finally had some time to look at this. I followed your suggestion and played around with it for a bit. 
For https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L476 I switched the calls so it goes rd high then cs high to match the GB CART code. I also did the same for https://github.com/sanni/cartreader/blob/master/Cart_Reader/GB.ino#L432 where I added the cs toggles and kept the same order consistency as the GB CART code. Which was cs low, rd low, (other code) rd high, cs high.
Fairly certain I also I kept the nops that totaled 2000ns as well. Although looking at his code he appears to only have 2, not sure if they equal different delays though.

I think we're getting closer to the answer and I made sure that the toggles looked identical for both calls (The comments use pull and toggle interchangeably but the code below them looks the same.) Finally I think it's worth noting at I check with a copy of Pokémon Red every so often to see if I can still read regular carts, and at this point it's not detecting it anymore. EDIT: (This was a mistake it totally does read Pokémon Red still, I'm just tired and put the cart in backwards lol), Here is my code for the two functions, also realized that I added nops for the write functions too, not sure if that was needed. I'll paste the two read functions I edited here. Final Edit: Forgot to mention the cart reader recognizes the GameBoy Camera but not the rom size or the sram size.

```
byte readByte_GB(word myAddress) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull CS(PH3) LOW
  PORTH &= ~(1 << 3);
  // Switch RD(PH6) to LOW
  PORTH &= ~(1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Read
  byte tempByte = PINC;

  // Switch and RD(PH6) to HIGH
  PORTH |= (1 << 6);
    // Pull CS(PH3) HIGH
  PORTH |= (1 << 3);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  return tempByte;
}
```



```
byte readByteSRAM_GB(word myAddress) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull CS(PH3) LOW
  PORTH &= ~(1 << 3);
  // Pull RD(PH6) LOW
  PORTH &= ~(1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Read
  byte tempByte = PINC;

  // Pull RD(PH6) HIGH
  PORTH |=  (1 << 6);
  // Pull CS(PH3) HIGH
  PORTH |= (1 << 3);


  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  return tempByte;
}

```

**Comment by MikalWhaley** [comment]
Hey! Happy to say this is the last update! Somewhere along the lines of messing with this I was able to get this to read and dump the save. But the kicker is that it read with ONLY the nops as a change, and didn't require editing for the two read functions we were talking about above. I think the contacts on my Game Boy Camera must be getting real old at this point because I just happened to get it to read this one time out of the 10 I tried. All of my other games read on the first try just not this one. For brevity sake and anyone who is reading this with the same issue just add 32 nops which is 2000ns and be patient unlike me.


Edit: Yeah, I'm definitely the worst person to test this. I tried for an hour to get this to even read again and confirm that the nops were indeed the only thing needed to get everything to read. Either my contacts are almost dead or the battery is almost dead lol.

**Comment by avapug** [comment]
I was having trouble getting my HW4 to read the camera cart. I finally got it working by changing the code to this.

```
byte readByte_GB(word myAddress) {
  PORTF = myAddress & 0xFF;
  PORTK = (myAddress >> 8) & 0xFF;

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Pull CS(PH3) CLK(PH1)(for FRAM MOD) LOW
  PORTH &= ~((1 << 3) | (1 << 1));
  // Pull RD(PH6) LOW
  PORTH &= ~(1 << 6);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  // Read
  byte tempByte = PINC;

  // Pull RD(PH6) HIGH
  PORTH |= (1 << 6);
  // Pull CS(PH3) CLK(PH1)(for FRAM MOD) HIGH
  PORTH |= (1 << 3) | (1 << 1);

  __asm__("nop\n\t""nop\n\t""nop\n\t""nop\n\t");

  return tempByte;
}
```

Surprisingly I didn't need to add the extra nops.

**Comment by sanni** [comment]
Thanks, I have updated the code and made a readByteCLK_GB function out of the code you posted and am using it as the default for both ROM and SRAM/FRAM.
